### ISSUE TITLE
Update `FormBuilder#control()` to support `AbstractControlOptions` as 2nd parameter

### ISSUE THREAD

**Issue by Klinton90** [angular/angular#19163]
## I'm submitting a...
<pre><code>
[ ] Regression (a behavior that used to work and stopped working in a new release)
[x ] Bug report  
[ ] Feature request
[ ] Documentation issue or request
[ ] Support request 
</code></pre>

## Current behavior
In 5.x version `AbstractControlOptions` has been added as parameter to `FormControl` constructor. `FormBuilder#control()` method should create new control by provided configuration. Currently signature/input parameters do not match `FormControl`.

## Expected behavior
```
control(
     formState: any = null,
      validatorOrOpts?: ValidatorFn|ValidatorFn[]|AbstractControlOptions|null,
      asyncValidator?: AsyncValidatorFn|AsyncValidatorFn[]|null
): FormControl {
    return new FormControl(formState, validatorOrOpts, asyncValidator);
  }
```

## Minimal reproduction of the problem with instructions
```
this.fb.group({
    "input1": this.fb.control('', {updateOn: 'submit'})
});
```

## What is the motivation / use case for changing the behavior?
Just to keep code consistent

## Environment

Angular version: 5.0.0-beta6

## Other

It would be nice, if you could also update `FormBuilder#group()` to support `AbstractControlOptions` when simple json is provided.

**Comment by artuska** [comment]
I also have this issue.

```
let someObject = {
  foo: 'bar'
};

let group = this.formBuilder.group(someObject, {updateOn: 'blur'});
```

Here the FormBuilder just ignores passed options.

**Comment by vivainio** [comment]
What's the recommended workaround? Not to use FormBuilder if you need these?

**Comment by michael-letcher** [comment]
I've also noted that in the [Form Builder - Group](https://angular.io/api/forms/FormBuilder#group)  docs it mentions

> Valid keys for the extra parameter map are validator and asyncValidator.

But in the [FormControl](https://angular.io/api/forms/FormControl) docs it show that you can add the updateOn.

I assume this was missed when updateOn was added.

**Comment by michael-letcher** [comment]
@vivainio 

This works for me for now

```
const controlConfig: {} = {};

// Loop through all fields and set default values and if field is required
fields.forEach(field => {
     controlConfig[field.key] = [field.value || '', { updateOn: 'blur', validators: field.validators || [] }];
});

return this.fb.group(controlConfig, { updateOn: 'blur' });
```

What should work though, but doesn't currently:

```
const controlConfig: {} = {};

// Loop through all fields and set default values and if field is required
fields.forEach(field => {
  controlConfig[field.key] = [field.value || '', field.validators || []];
});

return new FormGroup(controlConfig, { updateOn: 'blur' });
```

**Comment by dcesiel** [comment]
I had no luck with the workaround, I will wait for this to be resolved before using updateOn.

**Comment by jotapsl** [comment]
+1 for this feature. It's very inconvenient that we need to use FormControl constructors instead of FormBuilder to set validation onBlur.

**Comment by Buttonsz** [comment]
+1 for this for consistency and ease of use.

**Comment by dthomas722** [comment]
+1 for this for consistency and ease of use as well.

For others looking for a work around until this is resolved, the code below works for us:

```
let myFormGroup = this.fb.group({
  control1: 'control 1 value',
  control2: 'control 2 value'
});

myFormGroup = new FormGroup(formGroupWithoutBlur.controls, { updateOn: 'blur' });
```

**Comment by vytautas-pranskunas-** [comment]
Same here

**Comment by gassmannT** [comment]
+ 1. Same here

**Comment by GaryBirges** [comment]
Is it the same issue if I want to have 2 validators like this:

t```
his.newTransactionForm = this.fb.group({
 amount:              ['10', {Validators: [Validators.required,  Validators.min(1)],  updateOn: 'blur' }]
})
```

**Comment by fralewsmi** [comment]
Would love to see this feature

**Comment by LayZeeDK** [comment]
Thank you, @dthomas722!

Just to clear up your snippet

```typescript
let myFormGroup = this.formBuilder.group({
  control1: 'control 1 value',
  control2: 'control 2 value'
});

myFormGroup = new FormGroup(myFormGroup.controls, { updateOn: 'blur' });
```

or
```typescript
const myFormGroup = new FormGroup(this.formBuilder.group({
  control1: 'control 1 value',
  control2: 'control 2 value'
}).controls, {
 updateOn: 'blur',
});
```

**Comment by cezarcruz** [comment]
will be great!!

**Comment by pkaushik23** [comment]
@LayZeeDK  does the snippet work for you? i have the following code but does not work.

```
private toOnBlurFormGroup(){
    let group: any = {};
    
    this.controlCollection.forEach(element => {
        group[element.name] = new FormControl();  
    });
    return new FormGroup(this.fb.group(group).controls, { updateOn: 'blur' })
  }
```

**Comment by pkaushik23** [comment]
@kara just to add more information to the bug, the solution as suggested by @LayZeeDK & @dthomas722 works when we use a default html control like <input> but the solution does not work when we have a custom form control being used.

**Comment by pkaushik23** [comment]
@kara just to add more information to the bug, the solution as suggested by @LayZeeDK & @dthomas722 works when we use a default html control like **input** but the solution _does not work_ when we have a **custom form control** being used. example below:

`<my-text-edit [formControlName]="controlInfo.name" ></my-text-edit>`

**_For some one looking for a workaround_** (Please correct me if my usage is incorrect for custom form control)

The work around is to not to have the *[formControlName]* on the template where custom control is called as shown above, but to have it inside the template of *myTextEditComponent*. Although this does not look nice because now i have to pass in the value by exposing  another *@Input* property which takes the value for the control. for example

template for **myTextEditComponent**

`<input type="text" style="border: solid blue" [formControlName]="controlName"  />`

**And** the first code snippet changing to 

```
<my-text-edit [controlName]="controlInfoProp.name" [controlData] = "controlInfoProp.data" >
</my-text-edit>
```
**UPDATE**
if we dont specify the *[formControlName]* on the usage of **\<my-text-edit\>**, then methods of **controlValueAccessor** are not called. There i ultimately ended up doing following

while calling **\<my-text-edit\>**

```
<div [formGroup]="formGroupRef">
    <my-text-edit [controlName]="controlInfoProp.name" [formGroupRef]="formGroupRef" [formControlName]="controlInfoProp.name" ></my-text-edit>
</div>
```

Template of **myTextEditComponent**
```
<div [formGroup]="formGroupRef">
    <input type="text" style="border: solid blue" [formControlName]="controlName" (input)="OnValChange($event.target.value)"/>
</div>
```
Thanks,

**Comment by AhHa45** [comment]
what's the status on this?

**Comment by LayZeeDK** [comment]
>@IgorMinar Assigning to you for public API approval.
>⸻@kara in [#24599](https://github.com/angular/angular/pull/24599#issuecomment-429152078)

It is coming! 🙌🤞🤞

**Comment by LayZeeDK** [comment]
PR #24599 code has been merged. `updateOn` options for `FormBuilder` will be there in Angular 7.1.

**Comment by LayZeeDK** [comment]
@pkaushik23: @kara delivered [the best talk on Angular Forms at AngularConnect 2017](https://www.youtube.com/watch?v=CD_t3m2WMM8) where she shows how to avoid the anti-pattern of passing form groups down the component tree by injecting `NgControl` in the constructor of your custom control component.

**Comment by LayZeeDK** [comment]
You can now try out this feature in `@angular/forms@7.1.0-beta.2`.

**Comment by LayZeeDK** [comment]
@GaryBirges Try out [your example](https://github.com/angular/angular/issues/19163#issuecomment-376154489) in this StackBlitz:
https://stackblitz.com/edit/angular-formbuilder-update-on?file=app%2Ftransaction-form%2Ftransaction-form.component.ts

**Comment by GoNode5** [comment]
@LayZeeDK nice, thank you, what about documentation? Why is the option a string? Not very clear what are valid available options.

**Comment by LayZeeDK** [comment]
#24599 was released with `@angular/forms@7.1.0` 🎉

@kara You can close this issue.

**Comment by kara** [comment]
@LayZeeDK Thanks for your help getting this fixed! Closing as done.
